Hugging Face Hub Integration! (#4193)
voxel51/fiftyone: docs/source/integrations/huggingface.rst

    view = dataset.sort_by_similarity("A photo of a dog", k=25)

    session = fo.launch_app(view)

.. _huggingface-hub:

Hugging Face Hub
________________

FiftyOne integrates with the `Hugging Face Hub <https://huggingface.co/docs/hub/index>`_
to allow you to push datasets to and load datasets from the Hub with ease. This
integration simplifies the process of sharing datasets with the machine learning 
and computer vision community, and allows you to easily access and work with
many of the most popular vision and multimodal datasets available!

.. _huggingface-hub-setup:

Setup
-----

To push datasets to and load datasets from the `Hugging Face Hub <https://huggingface.co/docs/hub/index>`_,
you will need the `Hugging Face Hub Python client <https://github.com/huggingface/huggingface_hub>`_,
which you can install via PyPI:

.. code-block:: shell

    pip install -U huggingface_hub

> To push a dataset to the Hub, and in many cases to even access a dataset on

Reviewer: ```suggestion
To push a dataset to the Hub, and in some cases, to access a dataset on
```

.. _huggingface-hub-push-dataset-basic:

Basic usage
^^^^^^^^^^^

The basic recipe for pushing a FiftyOne dataset to the Hub is just two lines of
code. As a starting point, let's use the example 
:ref:`Quickstart dataset <dataset-zoo-quickstart>` dataset from the 
:ref:`FiftyOne Dataset Zoo <dataset-zoo>`:

.. code-block:: python
    :linenos:

    import fiftyone as fo
    import fiftyone.zoo as foz

    dataset = foz.load_zoo_dataset("quickstart")

To push the dataset to the Hugging Face Hub, all you need to do is call
:func:`push_to_hub() <fiftyone.utils.huggingface.push_to_hub>` with the dataset
and the desired `repo_name`:

.. code-block:: python
    :linenos:

    import fiftyone.utils.huggingface as fouh

    fouh.push_to_hub(dataset, "my-quickstart-dataset")

When you run this code, a few things happen:

- The dataset and its media files are exported to a temporary directory and
  uploaded to the specified Hugging Face repo.
- A `fiftyone.yml` config file for the dataset is generated and uploaded to
  the repo, which contains all of the necessary information so that the dataset
  can be loaded with :func:`load_from_hub() <fiftyone.utils.huggingface.load_from_hub>`.

> - A Hugging Face `Dataset Card <https://huggingface.co/docs/hub/en/datasets-cards>`_

Reviewer: super nice!

  for the dataset is auto-generated, providing tags, metadata, license info, and
  a code snippet illustrating how to load the dataset from the hub.

Your dataset will be available on the Hub at the following URL:

.. code-block:: shell

    https://huggingface.co/datasets/<your-username-or-org-name>/my-quickstart-dataset

> Pushing a |DatasetView| to the Hub works in exactly the same way. For example,

Reviewer: This is a really nice feature, getting a bit ahead of myself but I think this would be very good to communicate about in comms for this integration.

-----------------------------

To load a dataset from the Hugging Face Hub, you can use the
:func:`load_from_hub() <fiftyone.utils.huggingface.load_from_hub>` function.
This function supports loading datasets in any of the 
:ref:`common formats <supported-import-formats>` supported by FiftyOne, as well
as image-based datasets stored via `Parquet <https://parquet.apache.org/>`_ files,
as is common with datasets from the
`datasets <https://huggingface.co/docs/datasets/en/index>`_ library which have
been uploaded to the Hugging Face Hub. Below, we will walk through all of the
ways you can load datasets from the Hub.

In its simplest usage, the :func:`load_from_hub() <fiftyone.utils.huggingface.load_from_hub>`
function only requires the `repo_id` of the dataset you want to load. For example,
to load the :ref:`private dataset <huggingface-hub-push-dataset-advanced>` that
we pushed to the Hub earlier, you can do the following:

.. code-block:: python
    :linenos:

    import fiftyone.utils.huggingface as fouh

    dataset = fouh.load_from_hub("<username-or-org>/my-private-dataset")

.. note::

    As long as you have an environment variable `HF_TOKEN` set with your Hugging

>     Face token (with read access), you can load gated datasets that you have

Reviewer: ```suggestion
    Face token (with read access), you can load private or gated datasets that you have
```

Reviewer: We have private and gated datasets that need approval/agreement of some terms before accessing. The token will work the same either way but this might be clearer for end users.

    name: magicbrush
    filepath: source_img
    additional_media_fields:
        target_img: target_img
    mask_fields: mask_img

Now, you can load the dataset using the local yaml config file:

.. code-block:: python
    :linenos:

    import fiftyone.utils.huggingface as fouh

    dataset = fouh.load_from_hub(
        "osunlp/MagicBrush",
        config_file="/path/to/magicbrush.yml",
        max_samples=1000
    )

    session = fo.launch_app(dataset)

**Customizing the Download Process**:

When loading datasets from the Hub, you can customize the download process by
specifying the `batch_size`, `num_workers`, and `overwrite` arguments. For
example, to download the `full_numbers` subset of the `Street View House Numbers
<https://huggingface.co/datasets/svhn>`_ dataset with a batch size of 50 and 4
workers, you can do the following:

.. code-block:: python
    :linenos:

    import fiftyone.utils.huggingface as fouh

    dataset = fouh.load_from_hub(
        "svhn",
        format="parquet",
        detection_fields="digits",
        subsets="full_numbers",
        max_samples=1000,
        batch_size=50,
        num_workers=4
    )

    session = fo.launch_app(dataset)

> **Loading Gated Datasets**:

Reviewer: ```suggestion
**Loading Private or Gated Datasets**:
```

> Like public datasets, you can also load gated datasets from the Hub, as long as

Reviewer: ```suggestion
Like public datasets, you can also load private or gated datasets from the Hub, as long as
```
